pkg/resourcemanager/webhook/podzoneaffinity/handler.go

```diff
@@ -0,0 +1,224 @@
+	}
+
+	pod := &corev1.Pod{}
+	if err := h.decoder.Decode(req, pod); err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	// Check conflicting and add required pod affinity terms.
+	handlePodAffinity(pod)
+
+	// If the concrete zone is already determined by Gardener, let the pod be scheduled only to nodes in that zone.
+	if err := handleNodeAffinity(ctx, h.client, pod); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	marshaledPod, err := json.Marshal(pod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+func handlePodAffinity(pod *corev1.Pod) {
+	affinityTerm := corev1.PodAffinityTerm{
+		// Use empty label selector to match any pods in the shoot control plane namespace.
+		LabelSelector: &metav1.LabelSelector{},
+		TopologyKey:   corev1.LabelTopologyZone,
+	}
+
+	// First remove potentially conflicting pod anti affinity terms that would forbid scheduling into a single zone.
+	removeConflictingPodAntiAffinityTerms(pod, affinityTerm)
+
+	// Handle required affinity to let the pod be scheduled to a specific zone.
+	handleZonePodAffinityTerm(pod, affinityTerm)
+}
+
+func removeConflictingPodAntiAffinityTerms(pod *corev1.Pod, affinityTerm corev1.PodAffinityTerm) {
+	if pod.Spec.Affinity == nil || pod.Spec.Affinity.PodAntiAffinity == nil {
+		return
+	}
+
+	// Remove pod anti-affinity rules with zone topology key.
+	remainingAntiAffinityTerms := make([]corev1.PodAffinityTerm, 0, len(pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution))
+	for _, term := range pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+		if equality.Semantic.DeepEqual(term, affinityTerm) {
+			continue
+		}
+		remainingAntiAffinityTerms = append(remainingAntiAffinityTerms, term)
+	}
+	pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution = remainingAntiAffinityTerms
+}
+
+func handleZonePodAffinityTerm(pod *corev1.Pod, affinityTerm corev1.PodAffinityTerm) {
+	if pod.Spec.Affinity == nil {
+		pod.Spec.Affinity = &corev1.Affinity{}
+	}
+	if pod.Spec.Affinity.PodAffinity == nil {
+		pod.Spec.Affinity.PodAffinity = &corev1.PodAffinity{}
+	}
+
+	var (
+		zoneTermExisting      bool
+		filteredAffinityTerms = make([]corev1.PodAffinityTerm, 0, len(pod.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution))
+	)
+
+	for _, term := range pod.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+		if equality.Semantic.DeepEqual(term, affinityTerm) {
+			zoneTermExisting = true
+		}
+
+		// If there is another affinity configured on zones, we assume that this will be conflicting.
+		if term.TopologyKey == corev1.LabelTopologyZone && !zoneTermExisting {
+			continue
+		}
+
+		filteredAffinityTerms = append(filteredAffinityTerms, term)
+	}
```

Comment:
Is there the option to harmonize this code with (more or less the same) code above by introducing a function operating on `[]PodAffinityTerm`?